src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,295 @@
+///                    in form of a pointer to enum KbdLocks.
+/// @param[out]  err  Location and description where error message is to
+///                   be saved.
+///
+/// @return -1 in case of error otherwise an integer specified for given OS.
+static mod_mask_t os_get_locks_status(mod_mask_t *const mods, Error *const err)
+{
+  const CFMutableDictionaryRef mdict = IOServiceMatching(kIOHIDSystemClass);
+  const io_service_t ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                                       (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    api_set_error(err, kErrorTypeException,
+                  "E5777: Unable to initiate the service.\n"
+                  "IOServiceGetMatchingService() Failed");
+    return -1;
+  }
+
+  io_connect_t ioc;
+  kern_return_t kr;
+  if ((kr = IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                          &ioc)) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    api_set_error(err, kErrorTypeException,
+                  "E5777: Unable to initiate the service.\n"
+                  "IOServiceOpen() Failed: %d", (int)kr);
+    return -1;
+  }
+  IOObjectRelease(ios);
+
+  bool stateCaps;
+  if ((kr = IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                      &stateCaps)) != kIOReturnSuccess) {
+    IOServiceClose(ioc);
+    api_set_error(err, kErrorTypeException,
+                  "E5778: Unable to query CapsLock state.\n"
+                  "IOHIDGetModifierLockState() Failed: %d", (int)kr);
+    return -1;
+  }
+
+  bool stateNums;
+  if ((kr = IOHIDGetModifierLockState(ioc, kIOHIDNumLockState,
+                                      &stateNums)) != kIOReturnSuccess) {
+    IOServiceClose(ioc);
+    api_set_error(err, kErrorTypeException,
+                  "E5779: Unable to query NumLock state:\n"
+                  "IOHIDGetModifierLockState() Failed: %d", (int)kr);
+    return -1;
+  }
+  if (stateCaps) {
+    *mods |= kCapsLock;
+  }
+  if (stateNums) {
+    *mods |= kNumLock;
+  }
+  IOServiceClose(ioc);
+  return kCapsLock | kNumLock;
+}
+#else
+/// Dummy function used in case of no known platform detected.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to enum KbdLocks.
+/// @param[out]  err  Location and description where error message is to
+///                   be saved.
+///
+/// @return 0 for all the cases.
+static mod_mask_t os_get_locks_status(mod_mask_t *const mods, Error *const err)
+{
+  *mods = 0;
+  return 0;
+}
+#endif
+
+/// Get information about Num/Caps/Scroll Lock state.
+///
+/// To be used in nvim_get_keyboard_mods_state() function.
+///
+/// @param[out]  dict  Pointer to dictionary where information about modifiers
+///                    is to be dumped.
+/// @param[out]  err  Location and description where error message is to
```

Comment:
It sounds like `err` contains description of location where error message is saved alongside with location itself. This is not the case, remove “and description”.